### PR TITLE
Kibana monitoring with Metricbeat and Filebeat as sidecars

### PR DESCRIPTION
**Pull request by thbkrkr**
Adds a new `monitoring` field to the Kibana resource to configure one or two different Elasticsearch references to set up stack monitoring with Metricbeat and log delivery with Filebeat. The referenced ES are used to send the data collected by the beats.

Very similar to #4528.

```yaml
apiVersion: kibana.k8s.elastic.co/v1
kind: Kibana
metadata:
  name: alpha
  namespace: production
spec:
  version: 7.14.0-SNAPSHOT
  monitoring:
    metrics:
      elasticsearchRefs: 
        - name: monitoring
          namespace: observability
    logs:
      elasticsearchRefs:
        - name: monitoring
          namespace: observability
```

<details>
<summary>YAML example for testing</summary>

```yaml
apiVersion: v1
kind: Namespace
metadata:
  name: production
---
apiVersion: v1
kind: Namespace
metadata:
  name: observability
---
#######################################################################
# Monitored Elasticsearch
#######################################################################
apiVersion: elasticsearch.k8s.elastic.co/v1
kind: Elasticsearch
metadata:
  name: alpha
  namespace: production
spec:
  version: 7.14.0-SNAPSHOT
  # --------------------- #
  monitoring:
    metrics:
      elasticsearchRefs:
        - name: monitoring
          namespace: observability
    logs:
      elasticsearchRefs:
        - name: monitoring
          namespace: observability
  # --------------------- #
  nodeSets:
  - name: master
    count: 2
    config:
      node.store.allow_mmap: false
    # podTemplate:
    #   spec:
    #     containers:
    #       - name: metricbeat
    #         resources:
    #           limits:
    #             memory: 150Mi
---
apiVersion: kibana.k8s.elastic.co/v1
kind: Kibana
metadata:
  name: alpha
  namespace: production
spec:
  version: 7.14.0-SNAPSHOT
  count: 1
  # --------------------- #
  monitoring:
    metrics:
      elasticsearchRefs: 
        - name: monitoring
          namespace: observability
    logs:
      elasticsearchRefs:
        - name: monitoring
          namespace: observability
  # --------------------- #
  elasticsearchRef:
    name: alpha
    namespace: production
---
#######################################################################
# Monitoring clusters
#######################################################################
apiVersion: elasticsearch.k8s.elastic.co/v1
kind: Elasticsearch
metadata:
  name: monitoring
  namespace: observability
spec:
  version: 7.14.0-SNAPSHOT
  nodeSets:
  - name: master
    count: 2
    config:
      node.store.allow_mmap: false
---
apiVersion: kibana.k8s.elastic.co/v1
kind: Kibana
metadata:
  name: monitoring
  namespace: observability
spec:
  version: 7.14.0-SNAPSHOT
  count: 1
  config:
    xpack.monitoring.ui.enabled: false
  elasticsearchRef:
    name: monitoring
    namespace: observability

```
</details>

Implementation notes:
- Add a new `KbMonitoringAssociation` to manage association between the "1 Kibana / n monitoring Elasticsearch clusters" association
- Update `HasMonitoring` to make the validations and the base Metricbeat config build generic

Limitations:
- Minimum supported Stack version is 7.14.0 (to benefit from ES_LOG_STYLE=file)
- With a custom Kibana image, you have to override the podTemplate to define the custom beat images.
- The monitored Kibana is not deployed until monitoring ES clusters are not ready yet or the association are not configured
- monitoring.[metrics|logs].elasticsearchRefs accepts only one Elasticsearch reference. It's a slice to future proof the API for Elastic agent

Relates to #4183.

### CODE REVIEW

File: pkg/controller/association/controller/kb_monitoring.go [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+	"github.com/elastic/cloud-on-k8s/pkg/utils/rbac"
+)
+
+const (
+	// KbAssociationLabelName marks resources created by this controller for easier retrieval.
+	KbAssociationLabelName = "kbassociation.k8s.elastic.co/name"
+	// KbAssociationLabelNamespace marks resources created by this controller for easier retrieval.
+	KbAssociationLabelNamespace = "kbassociation.k8s.elastic.co/namespace"
+	// KbAssociationLabelType marks the type of association.
+	KbAssociationLabelType = "kbassociation.k8s.elastic.co/type"
```
Comment (sebgl) [inline pkg/controller/association/controller/kb_monitoring.go:29]: I tried to homogenize those comments in all association controllers recently: https://github.com/elastic/cloud-on-k8s/blob/7c39ff040e1e15b3e9b41ceb9779fd9101e938cb/pkg/controller/association/controller/ent_es.go#L24-L32
let's try to stick with the same comment pattern?

File: pkg/controller/common/stackmon/config.go [hunk 2]
```diff
@@ -145,3 +152,60 @@ func mergeConfig(rawConfig string, config map[string]interface{}) ([]byte, error
...
+	Username string
+	Password string
+	IsSSL    bool
+	SSLPath  string
+	SSLMode  string
+}
+
+// buildMetricbeatBaseConfig builds the base configuration for Metricbeat with the Elasticsearch or Kibana modules used
+// to collect metrics for Stack Monitoring
+func buildMetricbeatBaseConfig(
```
Comment (sebgl) [inline pkg/controller/common/stackmon/config.go:169]: let's unit test this?

File: pkg/controller/common/stackmon/monitoring/monitoring.go [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+package monitoring
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	commonv1 "github.com/elastic/cloud-on-k8s/pkg/apis/common/v1"
+)
+
+// HasMonitoring is the interface implemented by an Elastic Stack application that supports Stack Monitoring ()
```
Comment (sebgl) [inline pkg/controller/common/stackmon/monitoring/monitoring.go:14]: `( )`?

File: pkg/controller/common/stackmon/monitoring/monitoring.go [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+func GetLogsAssociation(resource HasMonitoring) []commonv1.Association {
+	associations := make([]commonv1.Association, 0)
+	for _, ref := range resource.GetMonitoringLogsRefs() {
+		if ref.IsDefined() {
+			associations = append(associations, resource.MonitoringAssociation(ref))
+		}
+	}
+	return associations
+}
```
Comment (sebgl) [inline pkg/controller/common/stackmon/monitoring/monitoring.go:62]: Let's unit test those functions?

**Comment by sebgl** [comment]
I have checked out the branch and am trying with this manifest:

```yml
apiVersion: kibana.k8s.elastic.co/v1
kind: Kibana
metadata:
  name: kb
spec:
  version: 7.14.0-SNAPSHOT
  count: 1
  elasticsearchRef:
    name: quickstart
  monitoring:
    metrics:
      elasticsearchRefs: 
        - name: quickstart
    logs:
      elasticsearchRefs:
        - name: quickstart
---
apiVersion: elasticsearch.k8s.elastic.co/v1
kind: Elasticsearch
metadata:
  name: quickstart
spec:
  version: 7.14.0-SNAPSHOT
  nodeSets:
  - name: default
    count: 1
    config:
      node.store.allow_mmap: false
```

I'm getting this error in the logs:

```
	ERROR	manager.eck-operator.controller.kibana-controller	Reconciler error	{"service.version": "1.7.0-SNAPSHOT+7c474bed", "name": "kb", "namespace": "default", "error": "Secret \"quickstart-es-internal-users\" not found", "errorCauses": [{"error": "Secret \"quickstart-es-internal-users\" not found"}]}
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
	/Users/sebgl/work/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.9.2/pkg/internal/controller/controller.go:253
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
	/Users/sebgl/work/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.9.2/pkg/internal/controller/controller.go:214
```

And don't have stack monitoring metrics visible in Kibana :(
Am I doing something wrong?

File: pkg/controller/kibana/controller.go [hunk 5]
```diff
@@ -9,6 +9,19 @@ import (
...
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
 	kbv1 "github.com/elastic/cloud-on-k8s/pkg/apis/kibana/v1"
```
Comment (sebgl) [inline pkg/controller/kibana/controller.go:25]: nit: looks like an import format inconsistency

File: pkg/controller/kibana/network/ports.go [hunk 6]
```diff
@@ -0,0 +1,10 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package network
+
+const (
+	// HTTPPort is the (default) port used by Kibana
+	HTTPPort = 5601
```
Comment (sebgl) [inline pkg/controller/kibana/network/ports.go:9]: why did you move this to its own sub-package?

File: pkg/controller/kibana/stackmon/filebeat.yml [hunk 7]
```diff
@@ -0,0 +1,21 @@
+filebeat.modules:
+  # https://www.elastic.co/guide/en/beats/filebeat/current/filebeat-module-kibana.html
```
Comment (sebgl) [inline pkg/controller/kibana/stackmon/filebeat.yml:2]: let's link to a concrete version rather than `current`

File: pkg/controller/kibana/stackmon/kb_config.go [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+		if cfg.Data == nil {
+			cfg.Data = map[string]interface{}{}
+		}
+		cfg.Data[XPackSecurityAuditEnabled] = true
+		cfg.Data[LoggingAppendersJSONFileAppenderType] = "rolling-file"
+		cfg.Data[LoggingAppendersJSONFileAppenderFilename] = filepath.Join(kibanaLogsMountPath, kibanaLogFilename)
+		cfg.Data[LoggingAppendersJSONFileAppenderLayoutType] = "json"
+		cfg.Data[LoggingAppendersJSONFileAppenderPolicyType] = "size-limit"
+		cfg.Data[LoggingAppendersJSONFileAppenderPolicySize] = "50mb"
+		cfg.Data[LoggingRootAppenders] = []string{"default", "rolling-file"}
```
Comment (sebgl) [inline pkg/controller/kibana/stackmon/kb_config.go:46]: does this mean we log to both stdout + a file? (may be worth mentioning in a comment)

File: pkg/controller/kibana/stackmon/kb_config.go [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+	LoggingAppendersJSONFileAppenderPolicyType = "logging.appenders.rolling-file.policy.type"
+	LoggingAppendersJSONFileAppenderPolicySize = "logging.appenders.rolling-file.policy.size"
+	LoggingRootAppenders                       = "logging.root.appenders"
+)
+
+// MonitoringConfig returns the Kibana settings required to enable the collection of monitoring data and disk logging
+func MonitoringConfig(kb kbv1.Kibana) commonv1.Config {
+	cfg := commonv1.Config{}
+	if monitoring.IsMetricsDefined(&kb) {
+		if cfg.Data == nil {
```
Comment (sebgl) [inline pkg/controller/kibana/stackmon/kb_config.go:31]: cfg.data must be nil since we initialized it just above?

File: pkg/controller/kibana/stackmon/sidecar.go [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+func Filebeat(client k8s.Client, kb kbv1.Kibana) (stackmon.BeatSidecar, error) {
+	filebeat, err := stackmon.NewFileBeatSidecar(client, &kb, kb.Spec.Version, filebeatConfig, nil)
+	if err != nil {
+		return stackmon.BeatSidecar{}, err
+	}
+
+	return filebeat, nil
+}
+
+// WithMonitoring updates the Elasticsearch Pod template builder to deploy Metricbeat and Filebeat in sidecar containers
```
Comment (sebgl) [inline pkg/controller/kibana/stackmon/sidecar.go:60]: `the Kibana Pod template builder`?

File: test/e2e/test/checks/monitoring.go [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+	// 1 index must exist
+	if len(indices) != 1 {
+		return fmt.Errorf("expected [%d] index [%s], found [%d]", len(indices), indexPattern, 1)
+	}
+	docsCount, err := strconv.Atoi(indices[0].DocsCount)
+	if err != nil {
+		return err
+	}
+	// with at least 1 doc
+	if docsCount < 0 {
```
Comment (sebgl) [inline test/e2e/test/checks/monitoring.go:126]: `<=`?

**Review by sebgl — COMMENTED** [review]
It's great that you could factorize Kibana and Elasticsearch monitoring into a single package!
I couldn't make things work (see my other comment).

**Comment by thbkrkr** [comment]
> And don't have stack monitoring metrics visible in Kibana :(

I forgot to mention that you need to also enable monitoring on the Elasticsearch associated to the monitored Kibana to be able to take advantage of the Kibana Monitoring UI (https://www.elastic.co/guide/en/kibana/current/monitoring-data.html#monitoring-data).
If you just enable monitoring on Kibana, you will just have metrics and logs collection without the monitoring view.

I wondered if we should force or validate this and I think it should be left open because it can make sense to just be interested in metrics and logs collection without the monitoring view.

> Am I doing something wrong?

No, you just tested with an elasticsearchRef with no namespace and it revealed a bug. I've always tested with namespaces :facepalm:.

File: pkg/controller/association/controller/kb_monitoring.go (hunk 1 above)
Comment (thbkrkr) [reply pkg/controller/association/controller/kb_monitoring.go:29]: I will update comments in es_monitoring.go and reuse `KibanaAssociationLabelXX`constants defined in kibana_es.go instead of introducing new ones.

File: pkg/controller/kibana/network/ports.go (hunk 6 above)
Comment (thbkrkr) [reply pkg/controller/kibana/network/ports.go:9]: To reuse it without import cycle. Same for the kbnamer. Another option could be to move it in `pkg/apis/kibana/v1` but I replicated what we did for ES.

File: pkg/controller/kibana/stackmon/kb_config.go (hunk 9 above)
Comment (thbkrkr) [reply pkg/controller/kibana/stackmon/kb_config.go:31]: Good point :)

**Comment by sebgl** [comment]
> I forgot to mention that you need to also enable monitoring on the Elasticsearch associated to the monitored Kibana to be able to take advantage of the Kibana Monitoring UI (https://www.elastic.co/guide/en/kibana/current/monitoring-data.html#monitoring-data).
If you just enable monitoring on Kibana, you will just have metrics and logs collection without the monitoring view.
I wondered if we should force or validate this and I think it should be left open because it can make sense to just be interested in metrics and logs collection without the monitoring view.

Considering it's a "limitation" / behaviour of the stack and not of ECK I think it's fine to leave as is indeed 👍 

> Also, it will not work to monitor Kibana if it hasn't an associated Elasticsearch because we need to be able to retrieve the internal user dedicated for metrics collection to configure the Metricbeat kibana module. That on the other hand, I think, it makes more sense to be validated.

👍

File: pkg/apis/kibana/v1/webhook.go [hunk 12]
```diff
@@ -104,3 +107,12 @@ func checkSupportedVersion(k *Kibana) field.ErrorList {
 func checkNoDowngrade(prev, curr *Kibana) field.ErrorList {
 	return commonv1.CheckNoDowngrade(prev.Spec.Version, curr.Spec.Version)
 }
+
+func checkMonitoring(k *Kibana) field.ErrorList {
+	errs := validations.Validate(k, k.Spec.Version)
+	if monitoring.IsDefined(k) && !k.Spec.ElasticsearchRef.IsDefined() {
+		errs = append(errs, field.Invalid(field.NewPath("spec").Child("elasticsearchRef"), k.Spec.ElasticsearchRef,
+			"Monitoring does not support Kibana without associated Elasticsearch"))
```
Comment (sebgl) [inline pkg/apis/kibana/v1/webhook.go:115]: something a bit more explicit: `Kibana must be associated to an Elasticsearch cluster through elasticsearchRef in order to enable monitoring features`

Comment (sebgl) [reply pkg/apis/kibana/v1/webhook.go:115]: We should also call this function as part of the regular Kibana reconciliation?

File: pkg/controller/kibana/stackmon/sidecar.go [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+)
+
+func Metricbeat(client k8s.Client, kb kbv1.Kibana) (stackmon.BeatSidecar, error) {
+	if !kb.Spec.ElasticsearchRef.IsDefined() {
+		return stackmon.BeatSidecar{}, errors.New("monitoring does not support Kibana without associated Elasticsearch")
+	}
+	associatedEsNsn := kb.Spec.ElasticsearchRef.NamespacedName()
+	if associatedEsNsn.Namespace == "" {
+		associatedEsNsn.Namespace = kb.Namespace
+	}
```
Comment (sebgl) [inline pkg/controller/kibana/stackmon/sidecar.go:41]: I'm surprised we have that check for metricbeat but not for filebeat?

**Review by sebgl — COMMENTED** [review]
I did a few tests it seems to work fine 👍 
Left a few last comments, this is almost ready to 🚢

File: pkg/apis/kibana/v1/webhook.go (hunk 12 above)
Comment (thbkrkr) [reply pkg/apis/kibana/v1/webhook.go:115]: Yes and we already do it when we call `validate`:

https://github.com/elastic/cloud-on-k8s/blob/f52ca4700f56238c56186f0fc751ada563cf6dda/pkg/controller/kibana/controller.go#L187

File: pkg/controller/kibana/stackmon/sidecar.go (hunk 13 above)
Comment (thbkrkr) [reply pkg/controller/kibana/stackmon/sidecar.go:41]: The base Filebeat configuration is purely static while the base Metricbeat configuration is a template to inject the credentials of the monitored Elasticsearch/Kibana into it. So, for Metricbeat we need to retrieve the monitoring user/password using this namespaced name. There is no this need for Filebeat.

**Review by sebgl — APPROVED** [review]
LGTM

**Comment by thbkrkr** [comment]
I've enable audit logs and configure Filebeat to collect an audit logs file but I forgot to configure the audit logger to write to a file.

**Comment by thbkrkr** [comment]
I find a small bug in `SetAssociationStatusMap` and I kinda messed up the E2E test during a last refactor. I fixed the two but there is one strange thing left, the metricbeat index is not populated with Kibana monitoring. I need to double-check if it's normal or not when I come back on the 19th.

I leave the PR open if you still want to check and test but I think it's ok to be merged.

**Comment by thbkrkr** [comment]
jenkins test this please

File: test/e2e/test/checks/monitoring.go [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+					return fmt.Errorf("pod %s not ready", pod.Name)
+				}
+			}
+			return nil
+		})}
+}
+
+func (c StackMonitoringChecks) CheckMetricbeatIndex() test.Step {
+	return test.Step{
+		Name: "Check that documents are indexed in one metricbeat-* index",
```
Comment (barkbay) [inline test/e2e/test/checks/monitoring.go:61]: I believe stack metrics are expected to be found in `.monitoring-kibana-` when `xpack.enabled` is set to `true`:

```
metricbeat-7.14.0-2021.07.12-000001 0 p STARTED  0   208b 10.55.208.22 test-kb-mon-metrics-lf6v-es-masterdata-1
metricbeat-7.14.0-2021.07.12-000001 0 r STARTED  0   208b 10.55.210.23 test-kb-mon-metrics-lf6v-es-masterdata-0
.monitoring-kibana-7-mb-2021.07.12  0 r STARTED  8 59.5kb 10.55.208.22 test-kb-mon-metrics-lf6v-es-masterdata-1
.monitoring-kibana-7-mb-2021.07.12  0 p STARTED  8 59.4kb 10.55.210.23 test-kb-mon-metrics-lf6v-es-masterdata-0
```

File: test/e2e/kb/stack_monitoring_test.go [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+	steps := func(k *test.K8sClient) test.StepList {
+		c := checks.StackMonitoringChecks{
+			MonitoredNsn: k8s.ExtractNamespacedName(&monitored.Kibana),
+			Metrics:      metrics,
+			Logs:         logs,
+			K:            k,
+		}
+		return test.StepList{
+			c.CheckBeatSidecars(),
+			//c.CheckMetricbeatIndex(), TODO: investigate if it's normal that there is no document in this index when es monitoring is off
```
Comment (barkbay) [inline test/e2e/kb/stack_monitoring_test.go:51]: See my other comment, I'm not sure to understand why enabling es monitoring would move the stack metrics from `.monitoring-*` to `metricbeat-*`. I'll try to investigate.

Comment (barkbay) [reply test/e2e/kb/stack_monitoring_test.go:51]: If I enable Elasticsearch monitoring I get the expected `.monitoring-es-` index, but there is also a `metricbeat-` index created:

```
.monitoring-kibana-7-mb-2021.07.12  0     p      STARTED  142 183.4kb 10.55.210.25 monitoring-es-default-1
.monitoring-kibana-7-mb-2021.07.12  0     r      STARTED  142   166kb 10.55.209.29 monitoring-es-default-2
.monitoring-es-7-mb-2021.07.12      0     p      STARTED  839     1mb 10.55.210.25 monitoring-es-default-1
.monitoring-es-7-mb-2021.07.12      0     r      STARTED  195 682.8kb 10.55.208.23 monitoring-es-default-0
metricbeat-7.14.0-2021.07.12-000001 0     r      STARTED   24 151.6kb 10.55.208.23 monitoring-es-default-0
metricbeat-7.14.0-2021.07.12-000001 0     p      STARTED   24 151.6kb 10.55.209.29 monitoring-es-default-2
```

Having a look at these 24 documents in `metricbeat-7.14.0-2021.07.12-000001` they are actually all errors which, I guess, happened during metricbeat startup:

| Time                        | _index                              | error.message                                                                                                                                                                                  |
|-----------------------------|-------------------------------------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| Jul 12, 2021 @ 13:28:09.677 | metricbeat-7.14.0-2021.07.12-000001 | error determining if connected Elasticsearch node is master: error making http request: Get "https://localhost:9200/_nodes/_local/nodes": dial tcp 127.0.0.1:9200: connect: connection refused |

I think `metricbeat-*` should not be used in e2e tests to assess that monitoring is working as expected.

Comment (barkbay) [reply test/e2e/kb/stack_monitoring_test.go]: Fixed in https://github.com/elastic/cloud-on-k8s/pull/4618/commits/b600503d7787fe374fea10f07a87d065d0c12856

**Comment by barkbay** [comment]
> If you just enable monitoring on Kibana, you will just have metrics and logs collection without the monitoring view.
> I wondered if we should force or validate this and I think it should be left open because it can make sense to just be interested in metrics and logs collection without the monitoring view.

I hit the same issue, also I was not able to get metrics in the monitoring view, I guess it is because metrics are stored in `.monitoring-kibana-7-mb-*` instead of `metricbeat-*`.
I would be tempted to say that it should not be an ECK specific issue, and things should be improved in Kibana: if there are some data in any of the monitoring indices user should be able to see them.

**Comment by barkbay** [comment]
jenkins test this please

```
16:02:02      integration.go:97: 
16:02:02          	Error Trace:	integration.go:97
16:02:02          	            				handler_integration_test.go:82
16:02:02          	Error:      	Received unexpected error:
16:02:02          	            	failed waiting for all runnables to end within grace period of 30s: context deadline exceeded
16:02:02          	Test:       	TestDynamicEnqueueRequest
```

**Comment by barkbay** [comment]
jenkins test this please

```
16:42:02      integration.go:97: 
16:42:02          	Error Trace:	integration.go:97
16:42:02          	            				handler_integration_test.go:82
16:42:02          	Error:      	Received unexpected error:
16:42:02          	            	failed waiting for all runnables to end within grace period of 30s: context deadline exceeded
16:42:02          	Test:       	TestDynamicEnqueueRequest
```

**Review by barkbay — APPROVED** [review]
LGTM 🚀
